### PR TITLE
[inductor] Add lowering for scatter_add

### PR DESCRIPTION
**Pull request by desertfire**
Summary: As title.
There is a remaining issue for the cpp backend.

### CODE REVIEW

File: torchinductor/lowering.py [hunk 1]
```diff
@@ -1305,6 +1305,55 @@ def fn(idx):
...
+            src_loader(idx),
+            ops.constant(0, x.get_dtype()),
+        )
+
+    result = clone(x)
+    scatter = ir.Scatter(
+        device=x.get_device(),
+        dtype=x.get_dtype(),
+        inner_fn=fn,
+        ranges=x.get_size(),
```
Comment (jansel) [inline torchinductor/lowering.py:1344]: Should ranges be `index.get_size()`?

If we are iterating over the ranges of `x` I would expect we need to broadcast `index` into the same shape as `x`.  I haven't looked up the semantics of scatter_add.

**Review by jansel — CHANGES_REQUESTED** [review]
What is the error we are getting on CPU?

File: torchinductor/lowering.py [hunk 2]
```diff
@@ -1305,6 +1305,55 @@ def fn(idx):
...
+                condition = boundary_check
+            else:
+                condition = ops.and_(condition, boundary_check)
+        return ops.where(
+            condition,
+            src_loader(idx),
+            ops.constant(0, x.get_dtype()),
+        )
+
+    result = clone(x)
```
[the remark this answers is not in the document]
Comment (desertfire) [reply torchinductor/lowering.py:1339]: I was thinking about doing that after seeing how `index_put` is done. Let me fix it.

My general question is how is in-place op supported in TorchInductor? I think at some point you mentioned there is a functionalization pass in FX and another one in TorchDynamo. Can you share some code pointers to those (if my memory was correct)? I am thinking to give a talk on that to our study group.

File: torchinductor/lowering.py (hunk 1 above)
Comment (desertfire) [reply torchinductor/lowering.py:1344]: https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_ explicitly says `index` and `src` do not broadcast.

Comment (jansel) [reply torchinductor/lowering.py:1344]: How many total adds are performed?  If it is `index.numel()` then this line should be `index.get_size()`.

File: torchinductor/lowering.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jansel) [reply torchinductor/lowering.py:1339]: Inductor supports codegen of mutation ops.  This is one example of it.  Another is input mutation, which happens in some benchmarks.

AOT Autograd does not support mutation, so for training we functionalize here:
https://github.com/pytorch/torchdynamo/blob/7e1d440939f1e84d3ea6b44c9ebdec892229e75d/torchdynamo/optimizations/normalize.py#L325
That will soon switch to dispatcher functionalization.

`scatter_add()` might be the product of functionalizing `scatter_add_()`.  Unfortunately, it doesn't really make sense to codegen a functional scatter.  So we de-functionalize it in order to generate code.

File: torchinductor/lowering.py (hunk 1 above)
Comment (desertfire) [reply torchinductor/lowering.py]: Yes, you are right. The cpu backend crash was a SEGFAULT caused by this. Fixed in the new commit.

There is still one problem remaining. When `index` only contains one element, the generated Triton code produces incorrect result. I vaguely remember seeing you have filed some related Triton issue. If not, I can file one.

**Review by jansel — APPROVED** [review]
